Disable compatibility layer when the WooCommerce Product Grid Block block and WooCommerce Single Product Block are used

gigitux opened the pull request:
This PR introduces a new filter called `woocommerce_disable_compatibility_layer` that allows disabling the compatibility layer.
At the same time, using the filter, the compatibility layer is disabled when Archive Product templates and Single Product template have the WooCommerce Product Grid Block.

This PR is blocked by https://github.com/woocommerce/woocommerce-blocks/pull/8442


<!-- Reference any related issues or PRs here -->

Fixes #8505 

### Testing
#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

##### Archive Templates

1. With the [Twenty Twenty-Three](https://wordpress.org/themes/twentytwentythree/) theme installed, go to `Appearance » Editor`.
2. Edit the Product Catalog template, adding the Products block above the WooCommerce Product Grid Block.
3. Save.
4. Go to the front end and verify that the `WooCommerce Product Grid Block` is visible.


##### Single Product Template 

1. With the [Twenty Twenty-Three](https://wordpress.org/themes/twentytwentythree/) theme installed, go to `Appearance » Editor`.
2. Edit the Product Catalog template, adding the `Product Image Gallery block` above the WooCommerce Product Grid Block.
3. Save.
4. Go to the front end and verify that the `WooCommerce Product Grid Block` is visible.

* [x] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Performance Impact

<!-- Please document any known performance impact (positive or negative) here. If negative, provide some rationale for why this is an okay tradeoff or how this will be addressed. -->

gigitux committed:
Merge branch 'add/compatibility-single-product-block' into 8505-the-woocommerce-product-grid-block-block-is-not-visible-on-the-front-end-when-the-products-block-is-on-the-same-page

gigitux committed:
disable compatibility layer via hook

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into 8505-the-woocommerce-product-grid-block-block-is-not-visible-on-the-front-end-when-the-products-block-is-on-the-same-page

gigitux committed:
update docs

gigitux committed:
generate the docs

gigitux committed:
Merge branch 'add/compatibility-single-product-block' of https://github.com/woocommerce/woocommerce-blocks into 8505-the-woocommerce-product-grid-block-block-is-not-visible-on-the-front-end-when-the-products-block-is-on-the-same-page

gigitux committed:
Merge branch 'add/compatibility-single-product-block' of https://github.com/woocommerce/woocommerce-blocks into 8505-the-woocommerce-product-grid-block-block-is-not-visible-on-the-front-end-when-the-products-block-is-on-the-same-page

gigitux committed:
Merge branch 'add/compatibility-single-product-block' of https://github.com/woocommerce/woocommerce-blocks into 8505-the-woocommerce-product-grid-block-block-is-not-visible-on-the-front-end-when-the-products-block-is-on-the-same-page

gigitux committed:
fix version

Aljullu commented on src/Templates/AbstractTemplateCompatibility.php:
```diff
@@ -0,0 +1,179 @@
+				/**
+				* Filters to disable the compatibility layer for the blockified templates.
+				*
+				* This hooks allows to disable the compatibility layer for the blockified.
```
```suggestion
				* This hook allows to disable the compatibility layer for the blockified templates.
```

Aljullu commented on src/BlockTemplatesController.php:
```diff
@@ -325,11 +326,22 @@ function( $template ) {
+					if ( ! is_admin() && BlockTemplateUtils::template_has_legacy_template_block( $template ) ) {
+						add_filter( 'woocommerce_disable_compatibility_layer', '__return_true' );
+						return $template;
+					}
```
Nit: this `if` clause is the same here and in line 339-342. I wonder if it would make sense to move it before their parent if clauses, in line 328, so we avoid repeating it.

Aljullu commented on src/Templates/AbstractTemplateCompatibility.php:
```diff
@@ -0,0 +1,179 @@
+					return $block_content;
+				}
+
+				$this->inject_hooks( $block_content, $block );
```
Should this value be returned as well?
```suggestion
				return $this->inject_hooks( $block_content, $block );
```

Aljullu reviewed (commented):
I can confirm this PR removes the compatibility later from the Products block and Single Product Template blocks when the Classic block exists in the page. However, for me it didn't fix https://github.com/woocommerce/woocommerce-blocks/issues/8505. Is this a known issue? I see #8550 fixes it, but I'm not sure to understand the relation between both PRs. :confused: 

I would also update the PR title because currently only makes reference to the Product Grid, but the changes introduced also work in the Single Product template.

Besides that, I just left some small comments below (some of them are applicable to #8550 as well).

gigitux committed:
Merge branch 'add/compatibility-single-product-block' of https://github.com/woocommerce/woocommerce-blocks into 8505-the-woocommerce-product-grid-block-block-is-not-visible-on-the-front-end-when-the-products-block-is-on-the-same-page

gigitux committed:
fix import

gigitux committed:
Merge branch 'add/compatibility-single-product-block' of https://github.com/woocommerce/woocommerce-blocks into 8505-the-woocommerce-product-grid-block-block-is-not-visible-on-the-front-end-when-the-products-block-is-on-the-same-page

gigitux committed:
fix code after merge

gigitux committed:
Merge branch 'add/compatibility-single-product-block' into 8505-the-woocommerce-product-grid-block-block-is-not-visible-on-the-front-end-when-the-products-block-is-on-the-same-page

gigitux commented:
> I can confirm this PR removes the compatibility later from the Products block and Single Product Template blocks when the Classic block exists in the page. However, for me it didn't fix #8505. Is this a known issue? I see #8550 fixes it, but I'm not sure to understand the relation between both PRs. 😕
> 
> I would also update the PR title because currently only makes reference to the Product Grid, but the changes introduced also work in the Single Product template.
> 
> Besides that, I just left some small comments below (some of them are applicable to #8550 as well).

Thanks for the review, @Aljullu! I updated the PR title.

Also, I refactored this PR after the work that I did on #8505. In the end, it is the same solution. Also, I changed the destination branch to [add/compatibility-single-product-block](https://github.com/woocommerce/woocommerce-blocks/tree/add/compatibility-single-product-block), so the diff is smaller. When you will approve this PR, I will merge #8442 in the trunk!

Aljullu reviewed (approved):
After your last commits, this PR only includes changes to the docs/comments, is that correct or I'm missing anything?

If so, LGTM, but then maybe we should check `Do not include in the Testing Notes`? :thinking:

gigitux commented:
Thanks for the review!

> After your last commits, this PR only includes changes to the docs/comments, is that correct or I'm missing anything?
> 

Yep, it is correct. We move these hooks from `BlockTemplatesCompatbility.php` to `AbstractTemplatesCompatbility.php`.

>  If so, LGTM, but then maybe we should check `Do not include in the Testing Notes`? 🤔

👍
